metagraph/src/annotation/binary_matrix/row_sparse_disk/row_sparse_disk.cpp

```diff
@@ -0,0 +1,142 @@
+#include "row_sparse_disk.hpp"
+
+#include "common/logger.hpp"
+#include "common/utils/file_utils.hpp"
+#include "common/ifstream_with_name_and_offset.hpp"
+#include "graph/representation/succinct/boss.hpp"
+
+
+namespace mtg {
+namespace annot {
+namespace binmat {
+
+using mtg::common::logger;
+
+bool RowSparseDisk::Impl::get(Row row, Column column) const {
+    SetBitPositions set_bits = get_row(row);
+    SetBitPositions::iterator v = std::lower_bound(set_bits.begin(), set_bits.end(), column);
+    return v != set_bits.end() && *v == column;
+}
+
+std::vector<BinaryMatrix::Row> RowSparseDisk::Impl::get_column(uint64_t num_rows,
+                                                               Column column) const {
+    std::vector<Row> result;
+    for (Row row = 0; row < num_rows; ++row) {
+        if (get(row, column))
+            result.push_back(row);
+    }
+    return result;
+}
+
+BinaryMatrix::SetBitPositions RowSparseDisk::Impl::get_row(Row row) const {
+    assert(boundary_[boundary_.size() - 1] == 1);
+    uint64_t start_idx = row == 0 ? 0 : boundary_.select1(row) + 1;
+    uint64_t end_idx = boundary_.next1(start_idx);
+    SetBitPositions result;
+    result.reserve(end_idx - start_idx);
+    // In each row, the first value in `set_bits_` stores the first set bit,
+    // and all next ones store deltas pos[i] - pos[i-1].
+    uint64_t last = 0;
+    for (uint64_t i = start_idx; i != end_idx; ++i) {
+        result.push_back(last += set_bits_[i - row]);
+    }
+    return result;
+}
+
+std::vector<BinaryMatrix::SetBitPositions>
+RowSparseDisk::Impl::get_rows(const std::vector<Row> &row_ids) const {
+    std::vector<SetBitPositions> rows(row_ids.size());
+
+    for (size_t i = 0; i < row_ids.size(); ++i) {
+        rows[i] = get_row(row_ids[i]);
+    }
+
+    return rows;
+}
+
+bool RowSparseDisk::load(std::istream &f) {
+    auto _f = dynamic_cast<mtg::common::IfstreamWithNameAndOffset *>(&f);
+    assert(_f);
+    try {
+        f.read(reinterpret_cast<char *>(&num_columns_), sizeof(uint64_t));
```

Comment:
please use the methods in `serialization.hpp` for this kind of IO

Reply:
Fixed